public/src/components/channelManagement/richTextEditor.tsx

```diff
@@ -0,0 +1,407 @@
+  const { to, empty } = useCurrentSelection();
+
+  const url = (useAttrs().link()?.href as string) ?? '';
+  const [href, setHref] = useState<string>(url);
+
+  const linkPositioner = useMemo(() => createMarkPositioner({ type: 'link' }), []);
+
+  const onRemove = useCallback(() => {
+    return chain
+      .removeLink()
+      .focus()
+      .run();
+  }, [chain]);
+
+  const updateReason = useUpdateReason();
+
+  useLayoutEffect(() => {
+    if (!isEditing) {
+      return;
+    }
+
+    if (updateReason.doc || updateReason.selection) {
+      setIsEditing(false);
+    }
+  }, [isEditing, setIsEditing, updateReason.doc, updateReason.selection]);
+
+  useEffect(() => {
+    setHref(url);
+  }, [url]);
+
+  const submitHref = useCallback(() => {
+    setIsEditing(false);
+    const range = linkShortcut ?? undefined;
+
+    if (href === '') {
+      chain.removeLink();
+    } else {
+      chain.updateLink({ href, auto: false }, range);
+    }
+    chain.focus(range?.to ?? to).run();
+  }, [setIsEditing, linkShortcut, chain, href, to]);
+
+  const cancelHref = useCallback(() => {
+    setIsEditing(false);
+  }, [setIsEditing]);
+
+  const clickEdit = useCallback(() => {
+    if (empty) {
+      chain.selectLink();
+    }
+
+    setIsEditing(true);
+  }, [chain, empty, setIsEditing]);
+
+  return useMemo(
+    () => ({
+      href,
+      setHref,
+      linkShortcut,
+      linkPositioner,
+      isEditing,
+      clickEdit,
+      onRemove,
+      submitHref,
+      cancelHref,
+    }),
+    [href, linkShortcut, linkPositioner, isEditing, clickEdit, onRemove, submitHref, cancelHref],
+  );
+}
+
+// ReMirror/ProseMirror FLOATING LINK MENU BUTTONS
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+const FloatingLinkToolbar = () => {
+  const {
+    isEditing,
+    linkPositioner,
+    clickEdit,
+    onRemove,
+    submitHref,
+    href,
+    setHref,
+    cancelHref,
+  } = useFloatingLinkState();
+  const active = useActive();
+  const activeLink = active.link();
+  const { empty } = useCurrentSelection();
+  const linkEditItems: ToolbarItemUnion[] = useMemo(
+    () => [
+      {
+        type: ComponentItem.ToolbarGroup,
+        label: 'Link',
+        items: activeLink
+          ? [
+              // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+              { type: ComponentItem.ToolbarButton, onClick: () => clickEdit(), icon: 'pencilLine' },
+              { type: ComponentItem.ToolbarButton, onClick: onRemove, icon: 'linkUnlink' },
+            ]
+          : [
+              // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+              { type: ComponentItem.ToolbarButton, onClick: () => clickEdit(), icon: 'link' },
+            ],
+      },
+    ],
+    [clickEdit, onRemove, activeLink],
+  );
+
+  const items: ToolbarItemUnion[] = useMemo(() => linkEditItems, [linkEditItems]);
```

Comment:
I realise this is taken from the example, but do you know why we need `items` here? It's just a copy of `linkEditItems`, which itself is memoized...

Reply:
I'll look into this and see if there's efficiencies to be made in the links-related code

Reply:
Looking at the code, they're using two floating toolbars, and they have a useMemo for each of them. One appears when text gets selected and offers the user a chance to create a link; the other shows up when the user clicks in an already-created link and offers them the chance to edit/delete the link.

`      <FloatingToolbar items={items} positioner="selection" placement="top" enabled={!isEditing} />
      <FloatingToolbar
        items={linkEditItems}
        positioner={linkPositioner}
        placement="bottom"
        enabled={!isEditing && empty}
      />
`